homeassistant/components/pglab/discovery.py

```diff
@@ -0,0 +1,292 @@
+"""Discovery PG LAB Electronics devices."""
+
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+import json
+from typing import TYPE_CHECKING, Any
+
+from pypglab.device import Device as PyPGLabDevice
+from pypglab.mqtt import Client as PyPGLabMqttClient
+
+from homeassistant.components.mqtt import ReceiveMessage
+from homeassistant.components.mqtt.subscription import (
+    EntitySubscription,
+    async_prepare_subscribe_topics,
+    async_subscribe_topics,
+    async_unsubscribe_topics,
+)
+from homeassistant.const import Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import device_registry as dr, entity_registry as er
+from homeassistant.helpers.device_registry import CONNECTION_NETWORK_MAC
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect,
+    async_dispatcher_send,
+)
+from homeassistant.helpers.entity import Entity
+
+from .const import DISCOVERY_TOPIC, DOMAIN, LOGGER
+
+if TYPE_CHECKING:
+    from . import PGLABConfigEntry
+
+# Supported platforms.
+PLATFORMS = [
+    Platform.SWITCH,
+]
+
+# Used to create a new component entity.
+CREATE_NEW_ENTITY = {
+    Platform.SWITCH: "pglab_create_new_entity_switch",
+}
+
+
+class PGLabDiscoveryError(Exception):
+    """Raised when a discovery has failed."""
+
+
+def get_device_id_from_discovery_topic(topic: str) -> str | None:
+    """From the discovery topic get the PG LAB Electronics device id."""
+
+    # The discovery topic has the following format "pglab/discovery/[Device ID]/config"
+    split_topic = topic.split("/", 5)
+
+    # Do a sanity check on the string.
+    if len(split_topic) != 4:
+        return None
+
+    if split_topic[3] != "config":
+        return None
+
+    return split_topic[2]
+
+
+class DiscoverDeviceInfo:
+    """Keeps information of the PGLab discovered device."""
+
+    def __init__(self, pglab_device: PyPGLabDevice) -> None:
+        """Initialize the device discovery info."""
+
+        # Hash string represents the devices actual configuration,
+        # it depends on the number of available relays and shutters.
+        # When the hash string changes the devices entities must be rebuilt.
+        self._hash = pglab_device.hash
+        self._entities: list[tuple[str, str]] = []
+
+    def add_entity_id(self, entity: Entity) -> None:
+        """Add the entity id."""
+
+        # Do some checking, be sure that the entity has a unique id.
+        if entity.unique_id:
+            self._entities.append((entity.platform.domain, entity.unique_id))
+        else:
+            raise PGLabDiscoveryError("Unexpected entity unique id")
```

Comment:
Please change like this, there's no need to check things in runtime which can't happen:
```suggestion
        # PGLabEntity always have unique IDs
        if TYPE_CHECKING:
            assert entity.unique_id is not None
        self._entities.append((entity.platform.domain, entity.unique_id))
```

Reply:
ok